dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/engine/SupplementaryDataProvider.java

```diff
@@ -42,53 +42,54 @@
 import org.hisp.dhis.organisationunit.OrganisationUnit;
 import org.hisp.dhis.organisationunit.OrganisationUnitGroupService;
 import org.hisp.dhis.programrule.ProgramRule;
+import org.hisp.dhis.rules.api.RuleSupplementaryData;
 import org.hisp.dhis.user.UserDetails;
 import org.springframework.stereotype.Component;
 
 @Component
 @RequiredArgsConstructor
 public class SupplementaryDataProvider {
-  private static final String USER = "USER";
-
-  private static final String REGEX =
+  private static final String ORG_UNIT_GROUP_REGEX =
       "d2:inOrgUnitGroup\\( *(([\\d/\\*\\+\\-%\\. ]+)|"
           + "( *'[^']*'))*+( *, *(([\\d/\\*\\+\\-%\\. ]+)|'[^']*'))*+ *\\)";
 
-  private static final Pattern PATTERN = Pattern.compile(REGEX);
+  private static final Pattern ORG_UNIT_GROUP_PATTERN = Pattern.compile(ORG_UNIT_GROUP_REGEX);
 
   @Nonnull private final OrganisationUnitGroupService organisationUnitGroupService;
 
-  public Map<String, List<String>> getSupplementaryData(
+  public RuleSupplementaryData getSupplementaryData(
       List<ProgramRule> programRules, UserDetails user) {
-    List<String> orgUnitGroups = new ArrayList<>();
 
-    for (ProgramRule programRule : programRules) {
-      Matcher matcher = PATTERN.matcher(StringUtils.defaultIfBlank(programRule.getCondition(), ""));
+    Map<String, List<String>> orgUnitGroupData = extractOrgUnitGroups(programRules);
+
+    return new RuleSupplementaryData(
+        user.getUserGroupIds().stream().toList(),
```

Comment:
will this always load all usergroups/roles regardless of whether they are even needed and which ones are needed by the expression?